heartbeat/monitors/active/icmp/loop.go

```diff
@@ -18,371 +18,15 @@
-		select {
-		case <-doneSignal:
-			ctx.Stop()
-
-		case r := <-ctx.result:
-			// ctx is removed from request tables automatically a response is
-			// received. No need to stop it.
-
-			// try to push RTT. The first result available will be reported
-			select {
-			case results <- r.packet.ts.Sub(ctx.ts):
-			default:
-			}
-		}
-	}
-
-	for !done {
-		var ctx *requestContext
-		ctx, err = l.sendEchoRequest(addr)
-		if err != nil {
-			close(doneSignal)
-			break
-		}
-		go awaitResponse(ctx)
-		requests++
-
-		select {
-		case <-toTimer.C:
-			// no response for any active request received. Finish loop
-			// and remove all requests from request table.
-			done = true
-			close(doneSignal)
-
-		case <-ticker.C:
-			// No response yet. Send another request with every tick
-
-		case rtt = <-results:
-			success = true
-
-			done = true
-			close(doneSignal)
-		}
-	}
-
-	if err != nil {
-		return 0, 0, err
-	}
-
-	if !success {
-		return 0, requests, timeoutError{}
-	}
-
-	return rtt, requests, nil
-}
-
-func (l *icmpLoop) sendEchoRequest(addr *net.IPAddr) (*requestContext, error) {
-	var conn *icmp.PacketConn
-	var proto int
-	var typ icmp.Type
-
-	if l == nil {
-		panic("icmp loop not initialized")
-	}
-
-	if isIPv4(addr.IP) {
-		conn = l.conn4
-		proto = protocolICMP
-		typ = ipv4.ICMPTypeEcho
-	} else if isIPv6(addr.IP) {
-		conn = l.conn6
-		proto = protocolIPv6ICMP
-		typ = ipv6.ICMPTypeEchoRequest
-	} else {
-		return nil, fmt.Errorf("%v is unknown ip address", addr)
-	}
-
-	id := requestID{
-		addr:  addr.String(),
-		proto: proto,
-		id:    rand.Intn(0xffff),
-		seq:   rand.Intn(0xffff),
-	}
-
-	ctx := &requestContext{
-		l:      l,
-		id:     id,
-		result: make(chan requestResult, 1),
-	}
-
-	l.mutex.Lock()
-	l.requests[id] = ctx
-	l.mutex.Unlock()
-
-	payloadBuf := make([]byte, 0, 8)
-	payload := bytes.NewBuffer(payloadBuf)
-	ts := time.Now()
-	binary.Write(payload, binary.BigEndian, ts.UnixNano())
-
-	msg := &icmp.Message{
-		Type: typ,
-		Body: &icmp.Echo{
-			ID:   id.id,
-			Seq:  id.seq,
-			Data: payload.Bytes(),
-		},
-	}
-	encoded, _ := msg.Marshal(nil)
-
-	_, err := conn.WriteTo(encoded, addr)
-	if err != nil {
-		return nil, err
-	}
-
-	ctx.ts = ts
-	return ctx, nil
-}
-
-func createListener(name, network string) *icmp.PacketConn {
-	conn, err := icmp.ListenPacket(network, "")
-
-	// XXX: need to check for conn == nil, as 'err != nil' seems always to be
-	//      true, even if error value itself is `nil`. Checking for conn suppresses
-	//      misleading log message.
-	if conn == nil && err != nil {
-		return nil
-	}
-	return conn
-}
-
-// timeoutError implements net.Error interface
-func (timeoutError) Error() string   { return "ping timeout" }
-func (timeoutError) Timeout() bool   { return true }
-func (timeoutError) Temporary() bool { return true }
-
-func (r *requestContext) Stop() {
-	r.l.mutex.Lock()
-	delete(r.l.requests, r.id)
-	r.l.mutex.Unlock()
+type ICMPLoop interface {
```

Comment:
This filename made more sense to hold the common interface that we can make a mock version of in tests.